// http2-express/dist/core.js
"use strict";

/* eslint-disable func-names */
/* eslint-disable no-underscore-dangle */
var EventEmitter = require('events');
var createHttp2Request = require('./http2Request');
var createHttp2Response = require('./http2Response');
var initHttp2Express = require('./initHttp2Express');
var _require = require('./util'),
  setPoweredBy = _require.setPoweredBy,
  isHttp2Request = _require.isHttp2Request;
var createHttp2Express = function createHttp2Express(express) {
  var application = express.application,
    request = express.request,
    response = express.response,
    Router = express.Router,
    query = express.query;
  var isExpress5 = !application.lazyrouter;
  if (isExpress5) {
    var originalInit = application.init;
    application.init = function init() {
      var _this = this;
      originalInit.call(this);
      // at this point we can use the router
      this.router.use(function (req, res, next) {
        setPoweredBy(_this, res);
        next();
      });
    };
  } else {
    application.lazyrouter = function lazyrouter() {
      if (!this._router) {
        this._router = new Router({
          caseSensitive: this.enabled('case sensitive routing'),
          strict: this.enabled('strict routing')
        });
        this._router.use(query(this.get('query parser fn')));
        this._router.use(initHttp2Express(this));
      }
    };
  }
  var oneshotRequestOverride = null;
  var oneshotResponseOverride = null;
  var _app = function app(req, res, next) {
    if (isExpress5 && isHttp2Request(req)) {
      // In Express 5, setting the prototype of req/res is done synchronously in the app.handle function,
      // so this is the right place to change the value of app.request/app.response.
      oneshotRequestOverride = _app.http2Request;
      oneshotResponseOverride = _app.http2Response;
    }
    _app.handle(req, res, next);
  };

  // Copying properties and descriptors from EventEmitter.prototype to app.
  Object.getOwnPropertyNames(EventEmitter.prototype).forEach(function (key) {
    var descriptor = Object.getOwnPropertyDescriptor(EventEmitter.prototype, key);
    if (descriptor) {
      Object.defineProperty(_app, key, descriptor);
    }
  });

  // Copy symbols from EventEmitter.prototype to app.
  Object.getOwnPropertySymbols(EventEmitter.prototype).forEach(function (symbol) {
    var descriptor = Object.getOwnPropertyDescriptor(EventEmitter.prototype, symbol);
    if (descriptor) {
      Object.defineProperty(_app, symbol, descriptor);
    }
  });

  // Copying properties from application to app.
  Object.getOwnPropertyNames(application).forEach(function (key) {
    var descriptor = Object.getOwnPropertyDescriptor(application, key);
    if (descriptor) {
      Object.defineProperty(_app, key, descriptor);
    }
  });

  // Copy symbols from application to app.
  Object.getOwnPropertySymbols(application).forEach(function (symbol) {
    var descriptor = Object.getOwnPropertyDescriptor(application, symbol);
    if (descriptor) {
      Object.defineProperty(_app, symbol, descriptor);
    }
  });
  response.push = function () {};
  _app._request = Object.create(request, {
    app: {
      configurable: true,
      enumerable: true,
      writable: true,
      value: _app
    }
  });
  _app._response = Object.create(response, {
    app: {
      configurable: true,
      enumerable: true,
      writable: true,
      value: _app
    }
  });
  var http2Request = createHttp2Request(request);
  var http2Response = createHttp2Response(response);
  _app.http2Request = Object.create(http2Request, {
    app: {
      configurable: true,
      enumerable: true,
      writable: true,
      value: _app
    }
  });
  _app.http2Response = Object.create(http2Response, {
    app: {
      configurable: true,
      enumerable: true,
      writable: true,
      value: _app
    }
  });

  // Expose the prototype that will get set on requests.
  Object.defineProperty(_app, 'request', {
    configurable: true,
    enumerable: true,
    get: function get() {
      if (oneshotRequestOverride) {
        var req = oneshotRequestOverride;
        oneshotRequestOverride = null;
        return req;
      }
      return _app._request;
    },
    set: function set(val) {
      _app._request = val;
    }
  });

  // Expose the prototype that will get set on responses.
  Object.defineProperty(_app, 'response', {
    configurable: true,
    enumerable: true,
    get: function get() {
      if (oneshotResponseOverride) {
        var res = oneshotResponseOverride;
        oneshotResponseOverride = null;
        return res;
      }
      return _app._response;
    },
    set: function set(val) {
      _app._response = val;
    }
  });
  _app.init();
  return _app;
};
module.exports = createHttp2Express;
